### PR TITLE
[XLA] Use simplified version of TF_ASSIGN_OR_RETURN for MSVC

### PR DESCRIPTION
**Pull request by rongjiecomputer**
This simplified version also works for GCC as well actually.

Split from #15310.

#15213

### CODE REVIEW

**Comment by tatatodd** [comment]
@rongjiecomputer What errors do you encounter without this patch?  It'd be nice to keep these special-cases due to compiler differences to a minimum.

File: tensorflow/compiler/xla/status_macros.h [hunk 1]
```diff
@@ -196,6 +196,22 @@ class StatusAdaptorForMacros {
...
 #define TF_STATUS_MACROS_CONCAT_IMPL(x, y) x##y
 
+#ifdef _MSC_VER
+
+#define TF_ASSIGN_OR_RETURN(lhs, rexpr)                                   \
+  TF_ASSIGN_OR_RETURN_IMPL(TF_STATUS_MACROS_CONCAT_NAME(_status_or_value, \
+                                                        __COUNTER__),     \
+                           lhs, rexpr)
+
+#define TF_ASSIGN_OR_RETURN_IMPL(statusor, lhs, rexpr) \
```
Comment (tatatodd) [inline tensorflow/compiler/xla/status_macros.h]: I think this is identical to TF_ASSIGN_OR_RETURN_IMPL for the non-msc case.  Let's refactor the code to minimize the differences.

File: tensorflow/compiler/xla/status_macros.h [hunk 2]
```diff
@@ -196,6 +196,22 @@ class StatusAdaptorForMacros {
 #define TF_STATUS_MACROS_CONCAT_NAME(x, y) TF_STATUS_MACROS_CONCAT_IMPL(x, y)
 #define TF_STATUS_MACROS_CONCAT_IMPL(x, y) x##y
 
+#ifdef _MSC_VER
+
+#define TF_ASSIGN_OR_RETURN(lhs, rexpr)                                   \
```
Comment (tatatodd) [inline tensorflow/compiler/xla/status_macros.h]: I think this is identical to TF_ASSIGN_OR_RETURN_IMPL_3 for the non-msc case.  Let's format the code to be identical, so that it's easier to see that they're the same.

**Review by tatatodd — CHANGES_REQUESTED** [review]

**Comment by rongjiecomputer** [comment]
Example:

```cpp
  TF_ASSIGN_OR_RETURN(a, foo(10));
```

In GCC, it will be expanded into:

```cpp
auto _status_or_value0 = (foo(10)); if (!_status_or_value0.ok()) { return _status_or_value0.status(); } a = std::move(_status_or_value0.ValueOrDie());
```

But in MSVC, it will be expanded into:

```cpp
(a, foo(10));
```

With warning C4003: not enough actual parameters for macro 'TF_STATUS_MACRO_GET_VARIADIC_IMPL'

You can see in https://stackoverflow.com/questions/47752606/msvc-preprocessor-bug , StackOverflower said the original implementation is not standard-compliant. It works in GCC because GCC allows `__VA_ARGS__` to be empty as a compiler extension. MSVC does not allow that.

The new implementation works in GCC as well. Can I just remove the old implementation and use the new one in all compilers?

**Comment by tatatodd** [comment]
@rongjiecomputer I see, thanks for the info.

I don't know the history behind why this macro was written the way it is.  And it's used all over the place, internally as well.

Perhaps the best thing to do is to try your suggestion.  Can you update your PR to use your new implementation for all compilers?

Once you're done, ping this issue.  If all the tests pass, we can try merging it.

**Comment by rongjiecomputer** [comment]
@tatatodd Done. It might surprise you that the diff is incredibly smaller after removing the old implementation.

**Comment by tatatodd** [comment]
I looked into this a bit more.

I think you're right that the original code was relying on non-standard empty `__VA_ARGS__` behavior.  Which means that another possible fix would be to add a dummy arg to the invocation of TF_STATUS_MACRO_GET_VARIADIC_IMPL:

```
// Trailing 0 is a dummy value.
#define TF_ASSIGN_OR_RETURN(...) \
  TF_STATUS_MACRO_GET_VARIADIC_IMPL(__VA_ARGS__, TF_ASSIGN_OR_RETURN_IMPL_3, \
                                    TF_ASSIGN_OR_RETURN_IMPL_2, 0) \
```

Note that the internal version that this is based on was written to deal with an optional 3rd argument to TF_ASSIGN_OR_RETURN.  But when it was ported to XLA, the optional 3rd argument was removed, but the underlying machinery was retained.  Except that it no longer accepts the optional 3rd argument at all.

Tagging @eliben since I believe he originally ported this bit of logic.

To summarize, I think your PR should work fine and makes things simpler.  So thanks, and let's get this merged!  :)

**Review by tatatodd — APPROVED** [review]

**Comment by tatatodd** [comment]
Jenkins, test this please.
